File: apple/HavenApp/MediaFilters.swift
import SwiftUI
import CoreImage
import CoreImage.CIFilterBuiltins
#if canImport(UIKit)
import UIKit
#else
import AppKit
#endif

/// Portable photo/video filters. The *look* of each filter is described by a `FilterSpec` — a
/// small set of platform-neutral parameters (white balance, tone, saturation, grain…). iOS
/// renders it with Core Image here; Android/Windows/Linux can render the exact same `FilterSpec`
/// with their own pipelines (GPUImage / shaders), so a photo looks identical on every client.
struct FilterSpec: Equatable, Codable {
    var temperatureK: Double = 6500   // white-balance target (neutral 6500K)
    var tint: Double = 0              // green ↔ magenta, −100…100
    var saturation: Double = 1        // 0 = grayscale, 1 = normal
    var contrast: Double = 1
    var brightness: Double = 0
    var vibrance: Double = 0          // −1…1, protects skin tones
    var highlights: Double = 1        // 0…1, pull highlights down
    var shadows: Double = 0           // −1…1, lift/deepen shadows
    var fade: Double = 0              // 0…1, lifted blacks (matte film look)
    var grain: Double = 0             // 0…1, film grain amount
    var vignette: Double = 0          // 0…1, corner darkening
    var monochrome: MonoTone? = nil   // non-nil = convert to a toned B&W

    enum MonoTone: String, Codable { case neutral, warm, cool }
}

/// The built-in filter set: nine Apple-Photos-style looks plus a realistic Kodak Gold film sim.
enum HavenFilter: String, CaseIterable, Identifiable {
    case original
    case vivid, vividWarm, vividCool
    case dramatic, dramaticWarm, dramaticCool
    case mono, silvertone, noir
    case kodakGold

    var id: String { rawValue }

    var title: String {
        switch self {
        case .original: return "Original"
        case .vivid: return "Vivid"
        case .vividWarm: return "Vivid Warm"
        case .vividCool: return "Vivid Cool"
        case .dramatic: return "Dramatic"
        case .dramaticWarm: return "Dramatic Warm"
        case .dramaticCool: return "Dramatic Cool"
        case .mono: return "Mono"
        case .silvertone: return "Silvertone"
        case .noir: return "Noir"
        case .kodakGold: return "Kodak Gold"
        }
    }

    var spec: FilterSpec {
        switch self {
        case .original:
            return FilterSpec()
        case .vivid:
            return FilterSpec(saturation: 1.22, contrast: 1.06, vibrance: 0.25)
        case .vividWarm:
            return FilterSpec(temperatureK: 5200, saturation: 1.2, contrast: 1.05, vibrance: 0.22)
        case .vividCool:
            return FilterSpec(temperatureK: 7800, saturation: 1.2, contrast: 1.05, vibrance: 0.22)
        case .dramatic:
            return FilterSpec(saturation: 0.92, contrast: 1.22, highlights: 0.85, shadows: -0.2)
        case .dramaticWarm:
            return FilterSpec(temperatureK: 5000, saturation: 0.95, contrast: 1.2, highlights: 0.85, shadows: -0.18)
        case .dramaticCool:
            return FilterSpec(temperatureK: 8200, saturation: 0.9, contrast: 1.2, highlights: 0.85, shadows: -0.2)
        case .mono:
            return FilterSpec(contrast: 1.05, monochrome: .neutral)
        case .silvertone:
            return FilterSpec(contrast: 1.1, monochrome: .warm)
        case .noir:
            return FilterSpec(contrast: 1.3, shadows: -0.3, monochrome: .cool)
        case .kodakGold:
            // Warm highlights, gently lifted/faded blacks, soft contrast, a touch of grain +
            // vignette — the classic consumer-film snapshot look.
            return FilterSpec(temperatureK: 5400, tint: 6, saturation: 1.08, contrast: 0.96,
                              vibrance: 0.12, highlights: 0.92, shadows: 0.08, fade: 0.12,
                              grain: 0.22, vignette: 0.18)
        }
    }
}

enum FilterEngine {
    static let context = CIContext(options: [.useSoftwareRenderer: false])

    /// Render a `FilterSpec` onto a PlatformImage. Returns the original on any failure.
    static func apply(_ filter: HavenFilter, to image: PlatformImage) -> PlatformImage {
        guard filter != .original, let cg = image.cgImage else { return image }
        let out = apply(filter.spec, to: CIImage(cgImage: cg))
        guard let result = context.createCGImage(out, from: out.extent) else { return image }
        #if canImport(UIKit)
        return PlatformImage(cgImage: result, scale: image.scale, orientation: image.imageOrientation)
        #else
        return PlatformImage(cgImage: result)
        #endif
    }

    /// The Core Image pipeline for a spec — also reused per-frame by the video compositor.
    static func apply(_ spec: FilterSpec, to input: CIImage) -> CIImage {
        var img = input

        if spec.temperatureK != 6500 || spec.tint != 0 {
            let f = CIFilter.temperatureAndTint()
            f.inputImage = img
            f.neutral = CIVector(x: 6500, y: 0)
            f.targetNeutral = CIVector(x: spec.temperatureK, y: spec.tint)
            img = f.outputImage ?? img
        }
        if spec.highlights != 1 || spec.shadows != 0 {
            let f = CIFilter.highlightShadowAdjust()
            f.inputImage = img
            f.highlightAmount = Float(spec.highlights)
            f.shadowAmount = Float(spec.shadows)
            img = f.outputImage ?? img
        }
        if spec.saturation != 1 || spec.contrast != 1 || spec.brightness != 0 {
            let f = CIFilter.colorControls()
            f.inputImage = img
            f.saturation = Float(spec.saturation)
            f.contrast = Float(spec.contrast)
            f.brightness = Float(spec.brightness)
            img = f.outputImage ?? img
        }
        if spec.vibrance != 0 {
            let f = CIFilter.vibrance(); f.inputImage = img; f.amount = Float(spec.vibrance)
            img = f.outputImage ?? img
        }
        if let tone = spec.monochrome {
            let f = CIFilter.colorMonochrome(); f.inputImage = img; f.intensity = 1
            switch tone {
            case .neutral: f.color = CIColor(red: 0.6, green: 0.6, blue: 0.6)
            case .warm: f.color = CIColor(red: 0.68, green: 0.62, blue: 0.5)
            case .cool: f.color = CIColor(red: 0.5, green: 0.55, blue: 0.62)
            }
            img = f.outputImage ?? img
        }
        if spec.fade > 0 {
            // Lift the black point toward a matte film base.
            let lift = spec.fade * 0.12
            let f = CIFilter.colorClamp()
            f.inputImage = img
            f.minComponents = CIVector(x: lift, y: lift, z: lift, w: 0)
            f.maxComponents = CIVector(x: 1, y: 1, z: 1, w: 1)
            img = f.outputImage ?? img
        }
        if spec.grain > 0 {
            img = addGrain(to: img, amount: spec.grain)
        }
        if spec.vignette > 0 {
            let f = CIFilter.vignette(); f.inputImage = img
            f.intensity = Float(spec.vignette); f.radius = Float(1.6)
            img = f.outputImage ?? img
        }
        return img.cropped(to: input.extent)
    }

    private static func addGrain(to image: CIImage, amount: Double) -> CIImage {
        let noise = CIFilter.randomGenerator().outputImage?
            .cropped(to: image.extent) ?? image
        let mono = CIFilter.colorControls()
        mono.inputImage = noise; mono.saturation = 0; mono.brightness = 0; mono.contrast = 1
        guard let grain = mono.outputImage else { return image }
        let blend = CIFilter.sourceOverCompositing()
        // Fade the grain to the requested strength.
        let faded = grain.applyingFilter("CIColorMatrix", parameters: [
            "inputAVector": CIVector(x: 0, y: 0, z: 0, w: amount * 0.5),
        ])
        blend.inputImage = faded; blend.backgroundImage = image
        return blend.outputImage ?? image
    }
}

/// A horizontal filter chooser with live thumbnails — drop it under a captured photo.
struct FilterStrip: View {
    let thumbnail: PlatformImage
    @Binding var selection: HavenFilter

    var body: some View {
        ScrollView(.horizontal, showsIndicators: false) {
            HStack(spacing: 12) {
                ForEach(HavenFilter.allCases) { f in
                    VStack(spacing: 6) {
                        Image(platformImage: FilterEngine.apply(f, to: thumbnail))
                            .resizable().scaledToFill()
                            .frame(width: 60, height: 60)
                            .clipShape(RoundedRectangle(cornerRadius: 10, style: .continuous))
                            .overlay(RoundedRectangle(cornerRadius: 10, style: .continuous)
                                .strokeBorder(selection == f ? HavenTheme.pink : .clear, lineWidth: 2.5))
                        Text(f.title).font(.caption2).lineLimit(1)
                            .foregroundStyle(selection == f ? HavenTheme.pink : .secondary)
                    }
                    .onTapGesture { withAnimation(HavenTheme.smooth) { selection = f } }
                }
            }
            .padding(.horizontal, 16).padding(.vertical, 8)
        }
    }
}

File: android/app/src/main/java/com/blaineam/haven/core/MediaFilters.kt
package com.blaineam.haven.core

/**
 * Portable photo/video filters — a direct port of the iOS `FilterSpec` / `HavenFilter` set
 * (apple/HavenApp/MediaFilters.swift). The *look* of each filter is a small set of platform-neutral
 * parameters; here we render that spec with a single GLSL fragment shader that is shared by the
 * photo path (offscreen GL), the video path (MediaCodec transcode), and the live preview — so a
 * photo/video looks the same on Android as it does on iOS.
 */
data class FilterSpec(
    val temperatureK: Double = 6500.0, // white-balance target (neutral 6500K)
    val tint: Double = 0.0,            // green ↔ magenta
    val saturation: Double = 1.0,      // 0 = grayscale, 1 = normal
    val contrast: Double = 1.0,
    val brightness: Double = 0.0,
    val vibrance: Double = 0.0,        // protects already-saturated tones
    val highlights: Double = 1.0,      // 0…1, pull highlights down
    val shadows: Double = 0.0,         // −1…1, lift/deepen shadows
    val fade: Double = 0.0,            // 0…1, lifted blacks (matte film look)
    val grain: Double = 0.0,           // 0…1, film grain
    val vignette: Double = 0.0,        // 0…1, corner darkening
    val monochrome: MonoTone? = null,
) {
    enum class MonoTone { NEUTRAL, WARM, COOL }
}

/** The built-in filter set — nine Apple-Photos-style looks plus a realistic Kodak Gold film sim. */
enum class HavenFilter(val title: String, val spec: FilterSpec) {
    ORIGINAL("Original", FilterSpec()),
    VIVID("Vivid", FilterSpec(saturation = 1.22, contrast = 1.06, vibrance = 0.25)),
    VIVID_WARM("Vivid Warm", FilterSpec(temperatureK = 5200.0, saturation = 1.2, contrast = 1.05, vibrance = 0.22)),
    VIVID_COOL("Vivid Cool", FilterSpec(temperatureK = 7800.0, saturation = 1.2, contrast = 1.05, vibrance = 0.22)),
    DRAMATIC("Dramatic", FilterSpec(saturation = 0.92, contrast = 1.22, highlights = 0.85, shadows = -0.2)),
    DRAMATIC_WARM("Dramatic Warm", FilterSpec(temperatureK = 5000.0, saturation = 0.95, contrast = 1.2, highlights = 0.85, shadows = -0.18)),
    DRAMATIC_COOL("Dramatic Cool", FilterSpec(temperatureK = 8200.0, saturation = 0.9, contrast = 1.2, highlights = 0.85, shadows = -0.2)),
    MONO("Mono", FilterSpec(contrast = 1.05, monochrome = FilterSpec.MonoTone.NEUTRAL)),
    SILVERTONE("Silvertone", FilterSpec(contrast = 1.1, monochrome = FilterSpec.MonoTone.WARM)),
    NOIR("Noir", FilterSpec(contrast = 1.3, shadows = -0.3, monochrome = FilterSpec.MonoTone.COOL)),
    // Warm highlights, gently lifted/faded blacks, soft contrast, a touch of grain + vignette.
    KODAK_GOLD("Kodak Gold", FilterSpec(temperatureK = 5400.0, tint = 6.0, saturation = 1.08, contrast = 0.96,
        vibrance = 0.12, highlights = 0.92, shadows = 0.08, fade = 0.12, grain = 0.22, vignette = 0.18)),
    ;

    companion object { val all = entries }
}

/** Builds the GLSL fragment shader for a spec, with the spec's constants baked in (no uniforms). */
object FilterShader {
    const val VERTEX = """
attribute vec4 aPosition;
attribute vec4 aTextureCoord;
varying highp vec2 vTextureCoord;
void main() {
    gl_Position = aPosition;
    vTextureCoord = aTextureCoord.xy;
}
"""

    /** A glsl float literal (always has a decimal point). */
    private fun g(v: Double): String {
        val s = "%.5f".format(v)
        return s
    }

    /** White balance (temperature/tint) → per-channel gains, matching iOS temperatureAndTint feel. */
    private fun wbGains(tempK: Double, tint: Double): Triple<Double, Double, Double> {
        val warm = (6500.0 - tempK) / 6500.0      // + warmer, − cooler
        val r = (1.0 + 0.40 * warm + 0.002 * tint).coerceIn(0.5, 1.6)
        val b = (1.0 - 0.40 * warm + 0.002 * tint).coerceIn(0.5, 1.6)
        val gr = (1.0 - 0.004 * tint).coerceIn(0.5, 1.6) // +tint → magenta (less green)
        return Triple(r, gr, b)
    }

    fun fragmentFor(spec: FilterSpec): String {
        val sb = StringBuilder()
        sb.append("precision mediump float;\n")
        sb.append("varying highp vec2 vTextureCoord;\n")
        sb.append("uniform lowp sampler2D sTexture;\n")
        sb.append("const vec3 L = vec3(0.299, 0.587, 0.114);\n")
        sb.append("void main() {\n")
        sb.append("    vec2 uv = vTextureCoord;\n")
        sb.append("    vec4 c = texture2D(sTexture, uv);\n")
        sb.append("    vec3 rgb = c.rgb;\n")

        val (rG, gG, bG) = wbGains(spec.temperatureK, spec.tint)
        if (spec.temperatureK != 6500.0 || spec.tint != 0.0)
            sb.append("    rgb *= vec3(${g(rG)}, ${g(gG)}, ${g(bG)});\n")

        if (spec.highlights != 1.0 || spec.shadows != 0.0) {
            sb.append("    { float l = dot(rgb, L);\n")
            sb.append("      rgb += ${g(spec.shadows)} * (1.0 - smoothstep(0.0, 0.5, l)) * 0.5;\n")
            sb.append("      rgb -= ${g(1.0 - spec.highlights)} * smoothstep(0.5, 1.0, l) * 0.5;\n")
            sb.append("      rgb = clamp(rgb, 0.0, 1.0); }\n")
        }
        if (spec.contrast != 1.0 || spec.brightness != 0.0)
            sb.append("    rgb = (rgb - 0.5) * ${g(spec.contrast)} + 0.5 + ${g(spec.brightness)};\n")

        val mono = spec.monochrome
        if (mono != null) {
            val tone = when (mono) {
                FilterSpec.MonoTone.NEUTRAL -> doubleArrayOf(0.6, 0.6, 0.6)
                FilterSpec.MonoTone.WARM -> doubleArrayOf(0.68, 0.62, 0.5)
                FilterSpec.MonoTone.COOL -> doubleArrayOf(0.5, 0.55, 0.62)
            }
            val luma = tone[0] * 0.299 + tone[1] * 0.587 + tone[2] * 0.114
            val n = if (luma > 0) 1.0 / luma else 1.0 // preserve overall brightness, just tint
            sb.append("    { float my = dot(clamp(rgb, 0.0, 1.0), L);\n")
            sb.append("      rgb = my * vec3(${g(tone[0] * n)}, ${g(tone[1] * n)}, ${g(tone[2] * n)}); }\n")
        } else {
            if (spec.saturation != 1.0)
                sb.append("    { float gy = dot(rgb, L); rgb = mix(vec3(gy), rgb, ${g(spec.saturation)}); }\n")
            if (spec.vibrance != 0.0) {
                sb.append("    { float mx = max(rgb.r, max(rgb.g, rgb.b));\n")
                sb.append("      float mn = min(rgb.r, min(rgb.g, rgb.b));\n")
                sb.append("      float sat = mx - mn; float gy = dot(rgb, L);\n")
                sb.append("      rgb = mix(vec3(gy), rgb, 1.0 + ${g(spec.vibrance)} * (1.0 - sat)); }\n")
            }
        }
        if (spec.fade > 0.0)
            sb.append("    rgb = max(rgb, vec3(${g(spec.fade * 0.12)}));\n")
        if (spec.grain > 0.0)
            sb.append("    { float n = fract(sin(dot(uv, vec2(12.9898, 78.233))) * 43758.5453);\n" +
                "      rgb += (n - 0.5) * ${g(spec.grain * 0.5)}; }\n")
        if (spec.vignette > 0.0)
            sb.append("    rgb *= 1.0 - ${g(spec.vignette)} * smoothstep(0.35, 0.85, distance(uv, vec2(0.5)));\n")

        sb.append("    gl_FragColor = vec4(clamp(rgb, 0.0, 1.0), c.a);\n")
        sb.append("}\n")
        return sb.toString()
    }
}
